File: CrypPlugins/PlayfairAnalyzer/Analysis/SolvePlayfair.cs
/*
   Copyright 2022 George Lasry, Nils Kopal, CrypTool 2 Team

   Licensed under the Apache License, Version 2.0 (the "License");
   you may not use this file except in compliance with the License.
   You may obtain a copy of the License at

       http://www.apache.org/licenses/LICENSE-2.0

   Unless required by applicable law or agreed to in writing, software
   distributed under the License is distributed on an "AS IS" BASIS,
   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
   See the License for the specific language governing permissions and
   limitations under the License.
*/
using PlayfairAnalysis.Common;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading;
using System.Threading.Tasks;

namespace PlayfairAnalysis
{
    public class SolvePlayfair
    {
        private static void Solve(int taskNumber, int saCycles, int innerRounds /* 200000 */, int multiplier/*1500*/, int[] cipherText, string crib, int minCribOffset, int maxCribOffset, Key simulationKey, AnalysisInstance instance, long[] evaluationsPerThread)
        {
            try
            {
                long totalCycles = 0;
                for (int cribOffset = minCribOffset; cribOffset <= maxCribOffset; cribOffset++)
                {
                    CancellationToken ct = instance.CancellationToken;
                    Utils utils = new Utils((int)DateTime.Now.Ticks + taskNumber * 100);  //Different seed for every thread
                    Transformations transformations = new Transformations(instance, utils);
                    SimulatedAnnealing simulatedAnnealing = new SimulatedAnnealing(utils);

                    long simulationOriginalScore = (simulationKey == null) ? long.MinValue : simulationKey.score;
                    Key currentKey = new Key(instance, utils);
                    Key newKey = new Key(instance, utils);
                    Key bestKey = new Key(instance, utils);
                    currentKey.SetCipher(cipherText);
                    newKey.SetCipher(cipherText);
                    bestKey.SetCipher(cipherText);
                    currentKey.SetCrib(crib, cribOffset);
                    newKey.SetCrib(crib, cribOffset);
                    bestKey.SetCrib(crib, cribOffset);

                    long serialCounter = 0;                    

                    for (int cycle = 0; cycle < saCycles || saCycles == 0; cycle++)
                    {
                        totalCycles++;
                        ct.ThrowIfCancellationRequested();
                        if (taskNumber == 0)
                        {
                            instance.CtAPI.UpdateProgress(totalCycles, saCycles * (maxCribOffset - minCribOffset + 1), evaluationsPerThread.Sum());
                        }

                        transformations.Randomize();
                        currentKey.Random();

                        long currentScore = currentKey.Eval();

                        bestKey.Copy(currentKey);
                        long bestScore = bestKey.Eval();
                        for (int innerRound = 0; innerRound < innerRounds; innerRound++)
                        {
                            ct.ThrowIfCancellationRequested();
                            transformations.Apply(currentKey, newKey, serialCounter++);

                            long newScore = newKey.Eval();
                            evaluationsPerThread[taskNumber]++;

                            if (simulatedAnnealing.AcceptHexaScore(newScore, currentScore, multiplier))
                            {
                                currentKey.Copy(newKey);
                                currentScore = newScore;

                                if (currentScore > bestScore)
                                {
                                    bestScore = currentScore;
                                    bestKey.Copy(currentKey);
                                    bestKey.Decrypt();
                                }
                            }
                        }

                        if (instance.CtBestList.ShouldPushResult(bestScore))
                        {
                            bestKey.AlignAlphabet();
                            (TimeSpan elapsed, long evaluations) = instance.Stats.EvaluationsSummary();
                            evaluations = evaluationsPerThread.Sum();
                            instance.CtBestList.PushResult(bestScore,
                                    bestKey.ToString(),
                                    bestKey.ToString(),
                                    Utils.GetString(bestKey.fullDecryption),
                                    elapsed,
                                    evaluations,
                                    instance.Stats.EvaluationsSummary() +
                                            $"[{bestKey.decryptionRemoveNullsLength}/{cipherText.Length}][Task: {taskNumber,2}][Mult.: {multiplier:N0}]");
                            if (currentScore == simulationOriginalScore || newKey.MatchesFullCrib())
                            {
                                instance.CtAPI.Printf("Key found");
                                instance.CtAPI.Goodbye();
                            }
                        }
                    }
                }
            }
            catch (OperationCanceledException)
            {
                //Let thread terminate
            }
        }

        public static long SolveMultithreaded(int[] cipherText, string cribString, int minCribOffset, int maxCribOffset, int threads, int cycles, Key simulationKey, AnalysisInstance instance)
        {
            const int innerRounds = 200_000;
            List<TaskCompletionSource<bool>> threadCompletions = new List<TaskCompletionSource<bool>>();
            Key simulationKey_ = simulationKey;
            long[] evaluationsPerThread = new long[threads];
            for (int t_ = 0; t_ < threads; t_++)
            {
                int t = t_;
                double factor = (cribString.Length > cipherText.Length / 2) ? 0.1 : 1.0;
                int multiplier = (int)(factor * 150_000) / cipherText.Length;

                TaskCompletionSource<bool> threadCompletion = new TaskCompletionSource<bool>();
                threadCompletions.Add(threadCompletion);
                new Thread(
                        delegate ()
                        {
                            try
                            {
                                Solve(t, cycles, innerRounds, multiplier, cipherText, cribString, minCribOffset, maxCribOffset, simulationKey_, instance, evaluationsPerThread);
                            }
                            finally
                            {
                                threadCompletion.SetResult(true);
                            }
                        }
                ).Start();
            }

            WatchThreadCompletions(threadCompletions, instance, evaluationsPerThread);

            return innerRounds * cycles * threads * (maxCribOffset - minCribOffset + 1);
        }

        private static async void WatchThreadCompletions(List<TaskCompletionSource<bool>> threadCompletions, AnalysisInstance instance, long[] evaluationsPerThread)
        {
            await Task.WhenAll(threadCompletions.Select(t => t.Task));

            instance.CtAPI.UpdateProgress(100, 100, evaluationsPerThread.Sum());
            //call "goodbye" after all threads completed to let CT2 know about it:
            instance.CtAPI.Goodbye();
        }
    }
}

File: Util/JavaCryptanalysis/src/playfair/SolvePlayfair.java
package playfair;

import common.*;

import java.util.Random;

public class SolvePlayfair {

    static void solve(int taskNumber, int saCycles, int innerRounds /* 200000 */, int multiplier/*1500*/, int[] cipherText, String crib, Key simulationKey) {

        long simulationOriginalScore = (simulationKey == null) ? Long.MIN_VALUE : simulationKey.score;
        Key currentKey = new Key();
        Key newKey = new Key();
        Key bestKey = new Key();
        currentKey.setCipher(cipherText);
        newKey.setCipher(cipherText);
        bestKey.setCipher(cipherText);
        currentKey.setCrib(crib);
        newKey.setCrib(crib);
        bestKey.setCrib(crib);

        long serialCounter = 0;

        for (int cycle = 0; cycle < saCycles || saCycles == 0; cycle++) {

            if (taskNumber == 0) {
                CtAPI.updateProgress(cycle, saCycles);
            }

            Transformations.randomize();
            currentKey.random();

            long currentScore = currentKey.eval();

            bestKey.copy(currentKey);
            long bestScore = bestKey.eval();
            for (int innerRound = 0; innerRound < innerRounds; innerRound++) {
                Transformations.apply(currentKey, newKey, serialCounter++);

                long newScore = newKey.eval();

                if (SimulatedAnnealing.acceptHexaScore(newScore, currentScore, multiplier)) {
                    currentKey.copy(newKey);
                    currentScore = newScore;

                    if (currentScore > bestScore) {
//                        currentKey.decrypt();
//                        long n8 = NGrams.eval8(currentKey.decryptionRemoveNulls, currentKey.decryptionRemoveNullsLength);
//
//                        if (CtBestList.shouldPushResult(n8)) {
//                            CtBestList.pushResult(n8,
//                                    currentKey.toString(),
//                                    currentKey.toString(),
//                                    Utils.getString(currentKey.fullDecryption),
//                                    Stats.evaluationsSummary() +
//                                            String.format("[%d/%d}[Task: %2d][Mult.: %,d]",
//                                                    currentKey.decryptionRemoveNullsLength, cipherText.length, taskNumber, multiplier));
//
//                        }
//
                        bestScore = currentScore;
                        bestKey.copy(currentKey);
                        bestKey.decrypt();
                    }

                }
            }

            if (CtBestList.shouldPushResult(bestScore)) {
                bestKey.alignAlphabet();
                CtBestList.pushResult(bestScore,
                        bestKey.toString(),
                        bestKey.toString(),
                        Utils.getString(bestKey.fullDecryption),
                        Stats.evaluationsSummary() +
                                String.format("[%d/%d}[Task: %2d][Mult.: %,d]",
                                        bestKey.decryptionRemoveNullsLength, cipherText.length, taskNumber, multiplier));
                if (currentScore == simulationOriginalScore || newKey.matchesFullCrib()) {
                    CtAPI.printf("Key found");
                    CtAPI.goodbye();
                }
            }
        }
    }

    static void solveMultithreaded(int[] cipherText, String cribString, int threads, int cycles, Key simulationKey) {
        Random r = new Random();
        final Key simulationKey_ = simulationKey;
        for (int t_ = 0; t_ < threads; t_++) {
            final int t = t_;
            final double factor = (cribString.length() > cipherText.length/2) ? 0.1 : 1.0;
            final int multiplier = (int) (factor* 150_000)/cipherText.length;

            new Thread(
                    () -> solve(t, cycles, 200_000, multiplier, cipherText, cribString, simulationKey_)
            ).start();
        }
    }
}
